public/app/plugins/datasource/cloudwatch/components/ConfigEditor.tsx

```diff
@@ -148,22 +133,26 @@ function useAuthenticationWarning(jsonData: CloudWatchJsonData) {
   }, [jsonData.authType, jsonData.database, jsonData.profile]);
 }
 
-function useDatasource(datasourceName: string, saved: boolean) {
+function useDatasource(props: Props) {
   const [datasource, setDatasource] = useState<CloudWatchDatasource>();
 
   useEffect(() => {
     // reload the datasource when it's saved
-    if (!saved) {
-      return;
+    if (!props.options.version || props.options.version < 2) {
```

Comment:
this code feels a bit off to me. If we haven't saved or we've saved only once then save? And then fetch it again asynchronously below which may or may not finish before this save?

do we need to save to the backend? Is it not enough to update it locally in state? I think I misunderstanding why we need anything here.

Also I think regardless we should probably add more tests for this so it's clear what's happening.

Reply:
This ConfigEditor needs to make async backend calls to load for example log groups. If the data source wasn't saved before, it needs to save it otherwise there's no backend to call. 

But I agree, the code was pretty horrible. Have done some refactoring and also written a few tests. Hopefully better now.